src/library/scala/collection/LinearSeq.scala

```diff
@@ -21,20 +21,33 @@ trait LinearSeqOps[+A, +CC[X] <: LinearSeq[X], +C <: LinearSeq[A] with LinearSeq
   // To be overridden in implementations:
   def isEmpty: Boolean
   def head: A
-  def tail: LinearSeq[A]
+  def tail: C
 
   def iterator: Iterator[A] = new LinearSeqIterator[A](toSeq)
```

Comment:
@NthPortal Should this perhaps check isEmpty and return Iterator.empty?

Reply:
check `isEmpty` - no. Check `knownSize == 0` - maybe

Reply:
@viktorklang I added a method `def knownIsEmpty: Boolean = stateEvaluated && isEmpty`, and optimised most lazy methods (including `iterator`) to check that first

Reply:
for some reason, I never realized this was about `LinearSeq` and not `LazyList`, so I've just fixed it